doc/Hammer-Use/Hammer-Config.rst

```diff
@@ -0,0 +1,193 @@
+
+```
+vlsi.tech.foobar65.bad_cells: ["NAND8X", "NOR8X"]
+```
+
+The solution is **meta variables**. This lets `hammer-config` know that instead of simply replacing the base variable, it should do a particular special action.
+
+In this case, we can use the `append` meta action:
+
+```
+vlsi.tech.foobar65.bad_cells: ["NAND8X", "NOR8X"]
+vlsi.tech.foobar65.bad_cells_meta: append
+```
+
+This will yield the desired result of `["NAND4X", "NOR4X", "NAND8X", "NOR8X"]` when `get_settings("vlsi.tech.foobar65.bad_cells")` is called in the end.
+
+Applying multiple meta actions
+------------------------------
+
+Multiple meta actions can be applied sequentially if the `_meta` variable is an array. Example:
+
+Layer 1:
+```
+foo.flash: yes
+```
+
+Layer 2 (located at /opt/foo):
+```
+foo.pipeline: "CELL_${foo.flash}.lef"
+foo.pipeline_meta: ['subst', 'prependlocal']
+```
+
+
+Result: `get_setting("foo.pipeline")` = `/opt/foo/CELL_yes.lef`
+
+Common meta actions
+-------------------
+
+* `append`: append the elements provided to the base list
+* `subst`: substitute variables into a string.
+
+Base:
+```
+foo.flash: yes
+```
+
+Meta:
+```
+foo.pipeline: "${foo.flash}man"
+foo.pipeline_meta: subst
+```
+
+Result: `get_setting("foo.flash")` = `yesman`
+
+* `lazysubst`: by default, variables are only substituted from previous configs. Using `lazysubst` allows us to deter the substitution until the very end.
+
+Example without `lazysubst`:
+
+Layer 1:
+```
+foo.flash: yes
+```
+
+Layer 2:
+```
+foo.pipeline: "${foo.flash}man"
+foo.pipeline_meta: subst
+```
+
+Layer 3:
+```
+foo.flash: no
+```
+
+Result: `get_setting("foo.flash")` = `yesman`
+
+Example with `lazysubst`:
+
+Layer 1:
+```
+foo.flash: yes
+```
+
+Layer 2:
+```
+foo.pipeline: "${foo.flash}man"
+foo.pipeline_meta: lazysubst
+```
+
+Layer 3:
+```
+foo.flash: no
+```
+
+Result: `get_setting("foo.flash")` = `noman`
+
+* `crossref` - directly reference another setting. Example:
+
+Layer 1:
+```
+foo.flash: yes
+```
+
+Layer 2:
+```
+foo.mob: "foo.flash"
+foo.mob_meta: crossref
+```
+
+Result: `get_setting("foo.mob")` = `yes`
+
+* `transclude` - transclude the given path. Example:
+
+Layer 1:
+```
+foo.bar: "/opt/foo/myfile.txt"
+foo.bar_meta: transclude
+```
+
+Result: `get_setting("foo.bar")` = `<contents of /opt/foo/myfile.txt>`
+
+* `prependlocal` - prepend the local path of this config file. Example:
+
+Layer 1 (located at /opt/foo):
+```
+foo.bar: "myfile.txt"
+foo.bar_meta: prependlocal
+```
+
+Result: `get_setting("foo.mob")` = `/opt/foo/myfile.txt`
+
+* `deepsubst` - like `subst` but descends into sub-elements. Example:
+
+Layer 1:
+```
+foo.bar: "123"
+```
+
+Layer 2:
+```
+foo.bar:
+  baz: "${foo.bar}45"
+  quux: "32${foo.bar}"
+foo.bar_meta: deepsubst
+```
+
+Result: `get_setting("foo.bar.baz")` = `12345` and `get_setting("foo.bar.baz")` = `32123`
+
+Reference
+---------
+
+For a more comprehensive view, please consult the `hammer_config` API documentation as well as
```

Comment:
I kind of wish there were more examples, but instead of immediately delving into code, perhaps list out some formulaic combinations of the `lazy`, `cross`, etc. modifiers for `append`, `ref`, etc. and give a small gist of what they might do?

Reply:
I agree - how do you think we could list it in a way that doesn't just turn into another list dump?

Reply:
Don't list out the actual combinations, rather something like this (though some don't quite fit in like transclude?)? You can use your judgment.

Base modifiers:
- lazy = leave til the end
- cross = <what is this?>
- ...

Base actions:
- append = add to end of list
- ref = look somewhere else
- ...

Most supported directives are a combination of these base modifiers and actions and the files below implement them:
...

Reply:
Turns out it's not so formulaic. Closing this.